File: perl5/package_skill.pl
#!/usr/bin/env perl
# package_skill.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/skill-creator/scripts/package_skill.py
# auch in: OpenClaw@gateway2:skills/skill-creator/scripts/package_skill.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use File::Spec;
use File::Basename;
use Cwd 'abs_path';
use File::Path qw(make_path);
use File::Find;
use Archive::Zip qw(:ERROR_CODES :CONSTANTS);

# Lade das Modul quick_validate.pm
require "./quick_validate.pm";

# Hilfsfunktion: Prüft, ob ein Pfad innerhalb eines anderen liegt
sub _is_within {
    my ($path, $root) = @_;
    my @path_parts = split('/', $path);
    my @root_parts = split('/', $root);

    # Entferne leere Elemente am Anfang
    shift @path_parts while @path_parts && $path_parts[0] eq '';
    shift @root_parts while @root_parts && $root_parts[0] eq '';

    # Vergleiche die Pfade
    for my $i (0 .. $#root_parts) {
        return 0 if $i > $#path_parts || $path_parts[$i] ne $root_parts[$i];
    }
    return 1;
}

# Paketiert einen Skill-Ordner in eine .skill-Datei
sub package_skill {
    my ($skill_path, $output_dir) = @_;

    # Normalisiere den Pfad
    $skill_path = abs_path($skill_path);

    # Prüfe, ob der Skill-Ordner existiert
    unless (-d $skill_path) {
        print "[ERROR] Skill folder not found or is not a directory: $skill_path\n";
        return undef;
    }

    # Prüfe, ob SKILL.md existiert
    my $skill_md = File::Spec->catfile($skill_path, "SKILL.md");
    unless (-f $skill_md) {
        print "[ERROR] SKILL.md not found in $skill_path\n";
        return undef;
    }

    # Validierung vor dem Packen
    print "Validating skill...\n";
    my ($valid, $message) = quick_validate::validate_skill($skill_path);
    unless ($valid) {
        print "[ERROR] Validation failed: $message\n";
        print "   Please fix the validation errors before packaging.\n";
        return undef;
    }
    print "[OK] $message\n\n";

    # Bestimme den Ausgabeort
    my $skill_name = basename($skill_path);
    my $output_path;
    if ($output_dir) {
        $output_path = abs_path($output_dir);
        make_path($output_path) unless -d $output_path;
    } else {
        $output_path = cwd();
    }

    my $skill_filename = File::Spec->catfile($output_path, "$skill_name.skill");

    my %EXCLUDED_DIRS = map { $_ => 1 } qw(.git .svn .hg __pycache__ node_modules);

    # Erstelle die .skill-Datei (ZIP-Format)
    eval {
        my $zip = Archive::Zip->new();

        # Sammle alle Dateien im Skill-Ordner
        my @files_to_add;
        find(
            sub {
                return if -l $_; # Symlinks überspringen
                my $rel_path = File::Spec->abs2rel($File::Find::name, $skill_path);
                my @parts = split('/', $rel_path);
                return if grep { $EXCLUDED_DIRS{$_} } @parts;

                if (-f $_) {
                    my $resolved_file = abs_path($_);
                    unless (_is_within($resolved_file, $skill_path)) {
                        die "[ERROR] File escapes skill root: $_\n";
                    }
                    if ($resolved_file eq abs_path($skill_filename)) {
                        print "[WARN] Skipping output archive: $_\n";
                        return;
                    }
                    push @files_to_add, [$File::Find::name, $rel_path];
                }
            },
            $skill_path
        );

        # Füge Dateien zum ZIP-Archiv hinzu
        for my $file_info (@files_to_add) {
            my ($full_path, $rel_path) = @$file_info;
            my $arcname = File::Spec->catfile($skill_name, $rel_path);
            $zip->addFile($full_path, $arcname);
            print "  Added: $arcname\n";
        }

        # Schreibe das ZIP-Archiv
        unless ($zip->writeToFileNamed($skill_filename) == AZ_OK) {
            die "[ERROR] Failed to write .skill file: $skill_filename\n";
        }

        print "\n[OK] Successfully packaged skill to: $skill_filename\n";
        return $skill_filename;
    };

    if ($@) {
        print $@;
        return undef;
    }

    return $skill_filename;
}

# Hauptprogramm
sub main {
    if (@ARGV < 1) {
        print "Usage: perl package_skill.pl <path/to/skill-folder> [output-directory]\n";
        print "\nExample:\n";
        print "  perl package_skill.pl skills/public/my-skill\n";
        print "  perl package_skill.pl skills/public/my-skill ./dist\n";
        exit 1;
    }

    my $skill_path = $ARGV[0];
    my $output_dir = $ARGV[1] // undef;

    print "Packaging skill: $skill_path\n";
    if ($output_dir) {
        print "   Output directory: $output_dir\n";
    }
    print "\n";

    my $result = package_skill($skill_path, $output_dir);

    exit($result ? 0 : 1);
}

main() if __FILE__ eq $0;

File: javascript/package_skill.js
#!/usr/bin/env node
// package_skill.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/skill-creator/scripts/package_skill.py
// auch in: OpenClaw@gateway2:skills/skill-creator/scripts/package_skill.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Skill Packager - Creates a distributable .skill file of a skill folder
 *
 * Usage:
 *     node utils/package_skill.js <path/to/skill-folder> [output-directory]
 *
 * Example:
 *     node utils/package_skill.js skills/public/my-skill
 *     node utils/package_skill.js skills/public/my-skill ./dist
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Import validate_skill function
function validate_skill(skillPath) {
    // Check if required files exist
    const requiredFiles = ['SKILL.md'];
    for (const file of requiredFiles) {
        if (!fs.existsSync(path.join(skillPath, file))) {
            return [false, `Missing required file: ${file}`];
        }
    }

    // Try to run skill validation script if it exists
    try {
        const validationScript = path.join(skillPath, 'validate.js');
        if (fs.existsSync(validationScript)) {
            execSync(`node "${validationScript}"`, { stdio: 'pipe' });
        }
        return [true, 'Validation passed'];
    } catch (error) {
        return [false, `Validation script failed: ${error.message}`];
    }
}

function _is_within(filePath, rootPath) {
    const relative = path.relative(rootPath, filePath);
    return relative && !relative.startsWith('..') && !path.isAbsolute(relative);
}

function package_skill(skillPath, outputDir = null) {
    // Resolve paths
    skillPath = path.resolve(skillPath);

    // Validate skill folder exists
    if (!fs.existsSync(skillPath)) {
        console.error(`[ERROR] Skill folder not found: ${skillPath}`);
        return null;
    }

    if (!fs.statSync(skillPath).isDirectory()) {
        console.error(`[ERROR] Path is not a directory: ${skillPath}`);
        return null;
    }

    // Validate SKILL.md exists
    const skillMd = path.join(skillPath, 'SKILL.md');
    if (!fs.existsSync(skillMd)) {
        console.error(`[ERROR] SKILL.md not found in ${skillPath}`);
        return null;
    }

    // Run validation before packaging
    console.log('Validating skill...');
    const [valid, message] = validate_skill(skillPath);
    if (!valid) {
        console.error(`[ERROR] Validation failed: ${message}`);
        console.error('   Please fix the validation errors before packaging.');
        return null;
    }
    console.log(`[OK] ${message}\n`);

    // Determine output location
    const skillName = path.basename(skillPath);
    let outputPath;
    if (outputDir) {
        outputPath = path.resolve(outputDir);
        fs.mkdirSync(outputPath, { recursive: true });
    } else {
        outputPath = process.cwd();
    }

    const skillFilename = path.join(outputPath, `${skillName}.skill`);

    const EXCLUDED_DIRS = new Set(['.git', '.svn', '.hg', '__pycache__', 'node_modules']);

    // Create the .skill file (zip format) using system zip command
    try {
        // Create temporary directory for packaging
        const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'skill-package-'));
        const tempSkillDir = path.join(tempDir, skillName);
        fs.mkdirSync(tempSkillDir);

        // Copy files to temporary directory
        function copyRecursiveSync(src, dest) {
            const stats = fs.statSync(src);
            if (stats.isDirectory()) {
                fs.mkdirSync(dest, { recursive: true });
                fs.readdirSync(src).forEach((childItemName) => {
                    // Skip excluded directories
                    if (EXCLUDED_DIRS.has(childItemName)) return;
                    
                    const srcPath = path.join(src, childItemName);
                    const destPath = path.join(dest, childItemName);
                    
                    // Security: never follow or package symlinks
                    if (fs.lstatSync(srcPath).isSymbolicLink()) {
                        console.warn(`[WARN] Skipping symlink: ${srcPath}`);
                        return;
                    }
                    
                    // Check if file escapes skill root
                    if (!_is_within(srcPath, skillPath)) {
                        console.error(`[ERROR] File escapes skill root: ${srcPath}`);
                        throw new Error('File escape detected');
                    }
                    
                    copyRecursiveSync(srcPath, destPath);
                });
            } else {
                // If output lives under skillPath, avoid writing archive into itself
                if (path.resolve(src) === path.resolve(skillFilename)) {
                    console.warn(`[WARN] Skipping output archive: ${src}`);
                    return;
                }
                
                fs.copyFileSync(src, dest);
                console.log(`  Added: ${path.join(skillName, path.relative(skillPath, src))}`);
            }
        }

        copyRecursiveSync(skillPath, tempSkillDir);

        // Create zip file
        const zipCommand = `cd "${tempDir}" && zip -r "${skillFilename}" "${skillName}"`;
        execSync(zipCommand, { stdio: 'inherit' });

        // Clean up temporary directory
        fs.rmSync(tempDir, { recursive: true, force: true });

        console.log(`\n[OK] Successfully packaged skill to: ${skillFilename}`);
        return skillFilename;

    } catch (error) {
        console.error(`[ERROR] Error creating .skill file: ${error.message}`);
        return null;
    }
}

function main() {
    const args = process.argv.slice(2);
    
    if (args.length < 1) {
        console.log('Usage: node utils/package_skill.js <path/to/skill-folder> [output-directory]');
        console.log('\nExample:');
        console.log('  node utils/package_skill.js skills/public/my-skill');
        console.log('  node utils/package_skill.js skills/public/my-skill ./dist');
        process.exit(1);
    }

    const skillPath = args[0];
    const outputDir = args[1];

    console.log(`Packaging skill: ${skillPath}`);
    if (outputDir) {
        console.log(`   Output directory: ${outputDir}`);
    }
    console.log();

    const result = package_skill(skillPath, outputDir);

    if (result) {
        process.exit(0);
    } else {
        process.exit(1);
    }
}

// Polyfill for os.tmpdir() since we're using it
const os = {
    tmpdir: () => process.env.TMPDIR || process.env.TEMP || process.env.TMP || '/tmp'
};

if (require.main === module) {
    main();
}

module.exports = { package_skill };
